### PR TITLE
enhancement: Improve error handling [CORE-2924]

### PR DESCRIPTION
**Pull request by joelvdavies**
Generally I tried to raise exceptions in the API functions, then in the actual CLI commands, or helper functions catch them where appropriate and use click to output nicer messages without the trace. Errors that are not directly caught will instead get the full trace, but these should only be for unforeseen cases where the logs are useful for development.

This PR:
- Removes cancel token from minio calls (CORE-2923)
- Removes the old raise_status parameter from session.py and all subsequent locations - the only place it was used was in dataset_upload.py for handling a 404 error separately
- Improves error handling so that:
    - `session.py` raises a separate error if one of the DAFNI API's returns an error message in its response
    - Handles a 404 in a way models, datasets and workflows now raise their own ResourceNotFoundError's which gets printed to the CLI when they are not found e.g. 
```
$ dafni get workflow sdfdsgfak
Unable to find a workflow with version id 'sdfdsgfak'
```
- Fixes uploading of new datasets
- Improves model validation (and dataset/workflow upload) to return all errors using the same code that handles errors in session.py
- Ensures datasets that fail to be uploaded delete the temporary bucket again after any exception/validation error to prevent a build up caused by repeated errors
- Fixes almost all remaining broken tests (apart from in utils.py)

Example of dataset errors:

image

Work for other tasks?:
---
- Workflow definition is not validated like models - it seems there is an endpoint for it so it could be implemented, but then again the same errors appear to come from the workflow upload anyway

### CODE REVIEW

File: dafni_cli/api/session.py [hunk 1]
```diff
@@ -327,15 +348,88 @@ def _authenticated_request(
...
+            if "error" in decoded_response:
+                error_message = f"Error: {decoded_response['error']}"
+            if "error_message" in decoded_response:
+                error_message = f"{error_message}, {decoded_response['error_message']}"
+            elif "errors" in decoded_response:
+                error_message = "The following errors were returned:"
+                for error in decoded_response["errors"]:
+                    error_message += f"\nError: {error}"
+            # Special case when uploading dataset metadata that's invalid
+            elif "metadata" in decoded_response:
```
Comment (roseDickinson) [inline dafni_cli/api/session.py:377]: I think this is actually a bug. The errors should be returned in an "errors" field not in  "metadata" - I'll add a task to JIRA to address this. Good spot

File: dafni_cli/api/session.py [hunk 2]
```diff
@@ -327,15 +348,88 @@ def _authenticated_request(
...
+
+        # If there is an error from DAFNI raise a DAFNI exception as well
+        # with more details, otherwise leave as any errors are HTTPError's
+        if error_message is None:
+            response.raise_for_status()
+        else:
+            try:
+                response.raise_for_status()
+            except HTTPError as err:
+                raise DAFNIError(error_message) from err
```
Comment (roseDickinson) [inline dafni_cli/api/session.py:423]: Could reduce nesting a little by doing 
```suggestion
        try:
            response.raise_for_status()
        except HTTPError as err:
            if error_message is None:
                raise err
            raise DAFNIError(error_message) from err
```

**Review by roseDickinson — APPROVED** [review]

File: dafni_cli/api/session.py (hunk 2 above)
Comment (joelvdavies) [reply dafni_cli/api/session.py:423]: Thanks!

File: test/datasets/test_dataset_upload.py [hunk 3]
```diff
@@ -1,203 +1,256 @@
...
+                call(f"Dataset ID: {details['datasetId']}"),
+                call(f"Version ID: {details['versionId']}"),
+                call(f"Metadata ID: {details['metadataId']}"),
+            ]
+        )
+
+    @patch("dafni_cli.datasets.dataset_upload.upload_files")
+    @patch("dafni_cli.datasets.dataset_upload.upload_metadata")
+    @patch("dafni_cli.datasets.dataset_upload.create_temp_bucket")
+    @patch("dafni_cli.datasets.dataset_upload.delete_temp_bucket")
```
[the remark this answers is not in the document]
Comment (joelvdavies) [reply test/datasets/test_dataset_upload.py:66]: Thanks, I will bear that in mind in future.

**Review by roseDickinson — APPROVED** [review]

**Pull request merged**